File: dataloom-frontend/src/api/visualizations.ts
/**
 * API functions for data visualization (chart suggestions + single charts).
 *
 * The backend computes aggregated chart data; every chart shares one
 * {@link ChartSpec} shape so the frontend maps `chart_type` to a renderer.
 * @module api/visualizations
 */
import client from "./client";

export type ChartType = "histogram" | "bar" | "line" | "area" | "scatter" | "pie";

/** Aggregation for a bar chart (matches the backend AggFunc enum). */
export type AggFunc = "sum" | "mean" | "median" | "min" | "max" | "count";

/** A single datum. x is a label, number, or ISO date; y is numeric. */
export interface ChartPoint {
  x: string | number | null;
  y: number | null;
}

/** A named set of points, optionally tinted with a color. */
export interface ChartSeries {
  name: string;
  data: ChartPoint[];
  color?: string | null;
}

/** Honesty flags about how the data was reduced for rendering. */
export interface ChartMeta {
  sampled?: boolean | null;
  truncated?: boolean | null;
  total_rows?: number | null;
  bins?: number | null;
}

/** Self-describing chart payload rendered as-is by the frontend. */
export interface ChartSpec {
  chart_type: ChartType;
  title: string;
  x_label: string;
  y_label: string;
  series: ChartSeries[];
  meta?: ChartMeta | null;
}

/** Parameters for {@link getChart}; required fields depend on `chart_type`. */
export interface ChartParams {
  chart_type: ChartType;
  /** Numeric column for a histogram. */
  column?: string;
  /** Grouping column for a bar/pie chart. */
  category?: string;
  /** Numeric value column for a bar/pie chart. */
  value?: string;
  /** X-axis column for a line/area/scatter chart. */
  x?: string;
  /** Y-axis column(s) for a line/area/scatter chart. */
  y?: string[];
  /** Optional color-grouping column for a scatter chart. */
  color?: string;
  /** Aggregation for a bar chart. */
  agg?: AggFunc;
  /** Bucket count for a histogram. */
  bins?: number;
}

/**
 * Fetch up to three charts recommended from the dataset's column shapes.
 * @param projectId - The project ID.
 * @returns Ready-to-render chart specs.
 */
export const getChartSuggestions = async (projectId: string): Promise<ChartSpec[]> => {
  const response = await client.get<{ suggestions: ChartSpec[] }>(
    `/projects/${projectId}/charts/suggest`,
  );
  return response.data.suggestions;
};

/**
 * Build a single chart from the project's dataset.
 * @param projectId - The project ID.
 * @param params - The chart configuration.
 * @returns The chart spec.
 */
export const getChart = async (projectId: string, params: ChartParams): Promise<ChartSpec> => {
  const response = await client.get<ChartSpec>(`/projects/${projectId}/charts`, {
    params,
    // FastAPI reads a repeated `y` (`y=a&y=b`); send it that way, not `y[]=`.
    paramsSerializer: { indexes: null },
  });
  return response.data;
};

File: dataloom-backend/app/api/endpoints/visualizations.py
"""Data visualization API endpoints.

Two read-only routes over a project's current working copy: auto-suggested charts
and a single configurable chart. Both return the self-describing ``ChartSpec``
shape (see ``schemas.ChartSpec``) — the backend aggregates, the frontend renders.

Column names are query parameters rather than path segments, matching the
profiling endpoints, so names with slashes or other reserved characters route
reliably across servers and proxies.
"""

import uuid

from fastapi import APIRouter, Depends, HTTPException, Query

from app import models, schemas
from app.api.dependencies import get_project_or_404, load_project_df
from app.services import visualization_service

router = APIRouter()


@router.get("/{project_id}/charts/suggest", response_model=schemas.ChartSuggestionsResponse)
async def suggest_charts(
    project_id: uuid.UUID,
    project: models.Project = Depends(get_project_or_404),
):
    """Return up to three charts recommended from the dataset's column shapes."""
    df = load_project_df(project)
    return {"suggestions": visualization_service.suggest_charts(df)}


@router.get("/{project_id}/charts", response_model=schemas.ChartSpec)
async def get_chart(
    project_id: uuid.UUID,
    chart_type: schemas.ChartType = Query(..., description="The kind of chart to build"),
    column: str | None = Query(None, description="Numeric column for a histogram"),
    category: str | None = Query(None, description="Grouping column for a bar/pie chart"),
    value: str | None = Query(None, description="Numeric value column for a bar/pie chart"),
    x: str | None = Query(None, description="X-axis column for a line/area/scatter chart"),
    y: list[str] = Query(default_factory=list, description="Y-axis column(s) for a line/area/scatter chart"),
    color: str | None = Query(None, description="Optional color-grouping column for a scatter chart"),
    agg: schemas.AggFunc = Query(schemas.AggFunc.sum, description="Aggregation for a bar chart"),
    bins: int = Query(visualization_service.DEFAULT_BINS, description="Bucket count for a histogram"),
    project: models.Project = Depends(get_project_or_404),
):
    """Build one chart from the project's dataset.

    Required parameters depend on ``chart_type``: histogram needs ``column``;
    bar/pie need ``category``; line/area/scatter need ``x`` and ``y``. Invalid
    combinations (missing params, or a column whose dtype the chart can't use)
    return 400; an unknown column returns 404.
    """
    df = load_project_df(project)
    try:
        return _dispatch(df, chart_type, column, category, value, x, y, color, agg, bins)
    except KeyError as e:
        raise HTTPException(status_code=404, detail=f"Column '{e.args[0]}' not found") from e
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e


def _dispatch(df, chart_type, column, category, value, x, y, color, agg, bins):
    """Route a chart request to the matching service builder, validating params."""
    if chart_type == schemas.ChartType.histogram:
        if not column:
            raise ValueError("histogram requires 'column'")
        return visualization_service.build_histogram(df, column, bins)

    if chart_type == schemas.ChartType.bar:
        if not category:
            raise ValueError("bar chart requires 'category'")
        return visualization_service.build_bar_chart(df, category, value, agg.value)

    if chart_type == schemas.ChartType.pie:
        if not category:
            raise ValueError("pie chart requires 'category'")
        return visualization_service.build_pie_chart(df, category, value)

    if chart_type in (schemas.ChartType.line, schemas.ChartType.area):
        if not x or not y:
            raise ValueError(f"{chart_type.value} chart requires 'x' and at least one 'y'")
        if chart_type == schemas.ChartType.line:
            return visualization_service.build_line_chart(df, x, y)
        return visualization_service.build_area_chart(df, x, y)

    # scatter
    if not x or len(y) != 1:
        raise ValueError("scatter chart requires 'x' and exactly one 'y'")
    return visualization_service.build_scatter_plot(df, x, y[0], color)
